text/0975-add-context-api.md

```diff
@@ -0,0 +1,148 @@
+
+This is where a context API would greatly improve the developer experience. Some
+examples of where context might be better than existing solutions.
+
+- A design system
+
+  In a component library, a `Card` component might accept a `backgroundColor`
+argument. The same component could then also expose this background color in its
+context state, making it available to all components nested inside it - no
+matter how deep.
+
+  Other library components, like buttons or text components, could then check what
+background color they're rendered on, and adjust their own styles accordingly,
+to make sure the text and background color contrasts are accessible, or just to
+make sure the button background matches the card background ("danger"-style
+buttons in "danger"-styled cards). These components could adjust their color no
+matter how deeply nested they are inside the card, without us having to
+explicitly pass the background color through multiple layers of components.
+
+- Chart and graphical components
+
+  In these scenarios there might be multiple layers being rendered, where yielding
+or passing certain arguments can become cumbersome. A `Graph` component could
+render an `Axis`, which would render `Tick`s and `TickLabel`s. The deeply-nested
+components will need access to the root `Graph`'s config to modify what and how
+they render.
+
+  While passing the config through many layers is an option, context could make
+this sort of code arguably easier to read, write and maintain.
+
+The feature has also been requested in this past, for example:
+- https://github.com/emberjs/rfcs/issues/775
+
+## Detailed design
+The details would still need to be worked out. The Glimmer VM already does a lot
+of the tree-tracking background work that would be necessary to pass context
+state through component layers.
+
+For example:
+- [DebugRenderTree](https://github.com/glimmerjs/glimmer-vm/blob/master/packages/%40glimmer/runtime/lib/debug-render-tree.ts)
+
+  This class is used by Ember Inspector, and it already tracks the hierarchy of
+components. We could do something similar to pass context state through
+component layers as they render.
+
+- [@glimmer/destroyable](https://github.com/glimmerjs/glimmer-vm/tree/master/packages/%40glimmer/destroyable)
+
+  When elements render, the Glimmer VM [builds parent/child associations](https://github.com/glimmerjs/glimmer-vm/blob/68d371bdccb41bc239b8f70d832e956ce6c349d8/packages/%40glimmer/destroyable/index.ts#L100-L101)
+in the destroyable module. Again, this is the sort of hierarchy tracking that
+would be needed for a context API.
+
+
+A very simple exploration into a `DebugRenderTree`-like context solution can be
+found at https://github.com/customerio/glimmer-vm/tree/provide-consume-context.
+This introduces a new class, which keeps track of "context provider" components,
+and exposes their state to any descendant components.
+
+The final implementation of context in Ember might be by using special
```

Comment:
Is there a way we could avoid string keys?

What would this look like in a template-only component?

Or gjs/gts?

Reply:
Great question! Template-only components are definitely an interesting use case to consider. One approach might be to also ship provider and consumer components, which could be invoked in any template, like:

```hbs
<ContextProvider @key="my-context-name" @value={{this.myValue}}>
</ContextProvider>
```

```hbs
<ContextConsumer @key="my-context-name" as |val|>
 {{val}}
</ContextConsumer>
```

Reply:
My general take is this should pretty intentionally mirror the service API, and if the way we declare services changes then the way we declare contexts also changes, but barring that the syntax is roughly identical